File: src/providers/supabase/rules/consistent-input-length-limits.ts
/**
 * supabase-consistent-input-length-limits (correctness)
 *
 * When several string fields are inserted together (e.g. all properties of
 * one `.insert(...)` call), and some of them are validated with an explicit
 * `.length > N` cap while at least one sibling field is only type-checked,
 * the uncapped field is usually the one the agent forgot — not a field that
 * was deliberately left unbounded. Flags the inconsistency, not "missing a
 * cap" in isolation (a single uncapped field with no capped siblings is out
 * of scope for this rule).
 */
import { chainObjectCall, memberPropName, resolvePropertyValueName, typeofStringCheckTarget } from '../utils.js';

interface VarValidation {
  typeofStringChecked: boolean;
  hasLengthCap: boolean;
}

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Sibling string fields inserted together should share the same length cap discipline',
      category: 'correctness',
      rationale:
        'A field with no length cap on an otherwise-validated, unauthenticated insert path lets a client persist arbitrarily large payloads repeatedly, growing storage with no bound. When sibling fields in the same insert already have explicit length caps, an uncapped field next to them is usually an oversight rather than an intentional choice.',
      docsUrl: 'https://supabase.com/docs/guides/database/tables',
      recommended: true,
    },
    messages: {
      inconsistentLengthLimit:
        'Field "{{field}}" has no length cap, but sibling field "{{cappedField}}" in the same insert does. Add a .length check for "{{field}}" too.',
    },
    schema: [],
  },
  create(context: any) {
    const validations = new Map<string, VarValidation>();

    function markValidation(name: string, key: keyof VarValidation) {
      let v = validations.get(name);
      if (!v) {
        v = { typeofStringChecked: false, hasLengthCap: false };
        validations.set(name, v);
      }
      v[key] = true;
    }

    function lengthCapTarget(node: any): string | undefined {
      if (node?.type !== 'BinaryExpression') return undefined;
      if (node.operator !== '>' && node.operator !== '>=') return undefined;
      const left = node.left;
      const right = node.right;
      if (left?.type !== 'MemberExpression') return undefined;
      if (left.property?.type !== 'Identifier' || left.property.name !== 'length') return undefined;
      if (left.object?.type !== 'Identifier') return undefined;
      if (right?.type !== 'Literal' || typeof right.value !== 'number') return undefined;
      return left.object.name;
    }

    return {
      BinaryExpression(node: any) {
        const typeofVar = typeofStringCheckTarget(node);
        if (typeofVar) markValidation(typeofVar, 'typeofStringChecked');

        const lengthVar = lengthCapTarget(node);
        if (lengthVar) markValidation(lengthVar, 'hasLengthCap');
      },

      CallExpression(node: any) {
        const prop = memberPropName(node);
        if (prop !== 'insert' && prop !== 'upsert') return;
        const objCall = chainObjectCall(node);
        if (!objCall || memberPropName(objCall) !== 'from') return;

        const arg = node.arguments?.[0];
        if (arg?.type !== 'ObjectExpression') return;

        const stringFields: { propNode: any; field: string; varName: string }[] = [];
        for (const p of arg.properties ?? []) {
          if (p?.type !== 'Property') continue;
          const field =
            p.key?.type === 'Identifier'
              ? p.key.name
              : p.key?.type === 'Literal'
                ? p.key.value
                : undefined;
          if (typeof field !== 'string') continue;
          const varName = resolvePropertyValueName(p);
          if (!varName) continue;
          const v = validations.get(varName);
          if (v?.typeofStringChecked) stringFields.push({ propNode: p, field, varName });
        }

        const capped = stringFields.filter((f) => validations.get(f.varName)?.hasLengthCap);
        const uncapped = stringFields.filter((f) => !validations.get(f.varName)?.hasLengthCap);
        if (capped.length === 0 || uncapped.length === 0) return;

        for (const f of uncapped) {
          context.report({
            node: f.propNode,
            messageId: 'inconsistentLengthLimit',
            data: { field: f.field, cappedField: capped[0].field },
          });
        }
      },
    };
  },
};

export const supabaseConsistentInputLengthLimitsRule = rule;
export default rule;

File: src/providers/supabase/rules/python/consistent_input_length_limits.py
"""supabase-consistent-input-length-limits

Parity with the JS rule: when several string fields inserted together are
validated with an `isinstance(x, str)` check, and some of them are also
capped with `len(x) > N`/`len(x) >= N` while at least one sibling is not,
the uncapped field is usually the one that was forgotten.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import (
    chain_object_call,
    is_supabase_table_call,
    isinstance_str_check_target,
    length_cap_target,
    member_prop_name,
    resolve_dict_value_name,
)

RULE_KEY = "supabase-consistent-input-length-limits"


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []
    # name -> {"typeof_checked": bool, "has_length_cap": bool}
    validations: dict[str, dict] = {}

    def mark(name: str, key: str) -> None:
        v = validations.setdefault(name, {"typeof_checked": False, "has_length_cap": False})
        v[key] = True

    for node in ast.walk(tree):
        target = isinstance_str_check_target(node)
        if target:
            mark(target, "typeof_checked")

        length_target = length_cap_target(node) if isinstance(node, ast.Compare) else None
        if length_target:
            mark(length_target, "has_length_cap")

    for node in ast.walk(tree):
        if not isinstance(node, ast.Call):
            continue
        prop = member_prop_name(node)
        if prop not in ("insert", "upsert"):
            continue
        obj_call = chain_object_call(node)
        if not obj_call or not is_supabase_table_call(obj_call):
            continue

        arg = node.args[0] if node.args else None
        if not isinstance(arg, ast.Dict):
            continue

        string_fields: list[tuple[ast.AST, str, str]] = []
        for k, v in zip(arg.keys, arg.values):
            field = k.value if isinstance(k, ast.Constant) and isinstance(k.value, str) else None
            if field is None:
                continue
            var_name = resolve_dict_value_name(v)
            if not var_name:
                continue
            validation = validations.get(var_name)
            if validation and validation["typeof_checked"]:
                string_fields.append((v, field, var_name))

        capped = [f for f in string_fields if validations.get(f[2], {}).get("has_length_cap")]
        uncapped = [f for f in string_fields if not validations.get(f[2], {}).get("has_length_cap")]
        if not capped or not uncapped:
            continue

        for value_node, field, _var_name in uncapped:
            line = getattr(value_node, "lineno", getattr(node, "lineno", 1)) or 1
            col = (getattr(value_node, "col_offset", getattr(node, "col_offset", 0)) or 0) + 1
            out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out
